### PR TITLE
Whitelisted smart contracts implementation

### PR DESCRIPTION
**Pull request by vpetrusenko**

### CODE REVIEW

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:blockchain_key][:desc] }
+          optional :description,
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:description][:desc] }
+          optional :address,
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:address][:desc] }
+          optional :state,
+                   values: { value: %w[active disabled], message: 'admin.whitelistedsmartcontract.invalid_state' },
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:state][:desc] }
+        end
+        post '/whitelisted_smart_contracts/update' do
```
Comment (mnaichuk) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: put '/whitelisted_smart_contracts'

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          requires :address,
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:address][:desc] }
+          optional :description,
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:description][:desc] }
+          optional :state,
+                   values: { value: %w[active disabled], message: 'admin.whitelistedsmartcontract.invalid_state' },
+                   default: 'active',
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:state][:desc] }
+        end
+        post '/whitelisted_smart_contracts/new' do
```
Comment (mnaichuk) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: post '/whitelisted_smart_contracts'

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        desc 'Get a whitelisted address.' do
+          success API::V2::Admin::Entities::WhitelistedSmartContract
+        end
+        params do
+          requires :id,
+                   type: { value: Integer, message: 'admin.whitelistedsmartcontract.non_integer_id' },
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:id][:desc] }
+        end
+        get '/whitelisted_smart_contracts/:id' do
```
Comment (mnaichuk) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: get '/whitelisted_smart_contract/:id'

File: app/models/wallet.rb [hunk 4]
```diff
@@ -68,10 +68,10 @@ class Wallet < ApplicationRecord
     end
   end
 
-  before_validation do
-    next unless address? && blockchain.blockchain_api.supports_cash_addr_format?
-    self.address = CashAddr::Converter.to_cash_address(address)
-  end
+  # before_validation do
```
Comment (mnaichuk) [inline app/models/wallet.rb]: this should be commented

File: app/models/whitelisted_smart_contract.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+class WhitelistedSmartContract < ApplicationRecord
+  # == Relationships ========================================================
+
+  belongs_to :blockchain, foreign_key: :blockchain_key, primary_key: :key
+
+  # == Validations ==========================================================
+
+  # validates :id, presence: true, uniqueness: true
+
+  validates :address, presence: true, uniqueness: { scope: :blockchain_key}
```
Comment (mnaichuk) [inline app/models/whitelisted_smart_contract.rb]: add empty space before last }

File: app/models/whitelisted_smart_contract.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  validates :state,  inclusion: { in: %w[active disabled] }
+
+  # == Scopes ===============================================================
+
+  scope :active, -> { where(state: active) }
+  scope :ordered, -> { order(kind: :asc) }
+
+  def explorer_url
+    blockchain.explorer_address.gsub('#{address}', address) if blockchain
```
Comment (mnaichuk) [inline app/models/whitelisted_smart_contract.rb]: as I remember you can just explorer_address from blockchain and FE will handle

File: db/migrate/20210128144535_create_whitelisted_smart_contracts.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+class CreateWhitelistedSmartContracts < ActiveRecord::Migration[5.2]
+  def change
+    create_table :whitelisted_smart_contracts do |t|
+      t.string  :description,    default: ''
+      t.string  :address,        default: '', null: false
```
Comment (mnaichuk) [inline db/migrate/20210128144535_create_whitelisted_smart_contracts.rb]: we don't need defaults here

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -90,7 +90,7 @@
     KlineDB.stubs(:kline).returns([])
     Currency.any_instance.stubs(:price).returns(1.to_d)
     %w[eth-kovan eth-rinkeby btc-testnet].each { |blockchain| FactoryBot.create(:blockchain, blockchain) }
-    %i[usd eur btc eth trst ring].each { |ccy| FactoryBot.create(:currency, ccy) }
+    %i[usd eur btc eth trst ring tom].each { |ccy| FactoryBot.create(:currency, ccy) }
```
Comment (mnaichuk) [inline spec/spec_helper.rb]: do not add more factories here you can create them if needed in spec

File: lib/peatio/ethereum/blockchain.rb [hunk 9]
```diff
@@ -39,9 +40,16 @@ def fetch_block!(block_number)
...
+            address = @whitelisted_addresses.include?(normalize_address(tx.fetch('to'))) &&
+                        normalize_address(tx.fetch('to'))
+          else
+            address = normalize_address(tx.fetch('to'))
+          end
+
           next if @erc20.find do |c|
             # Check `to` and `input` options to find erc-20 smart contract contract 
-            c.dig(:options, :erc20_contract_address) == normalize_address(tx.fetch('to')) ||
+            c.dig(:options, :erc20_contract_address) == address ||
```
Comment (mnaichuk) [inline lib/peatio/ethereum/blockchain.rb:47]: maybe more clear will be check if `normalize_address(tx.fetch('to'))` in @whitelisted_addresses ?

**Review by mnaichuk — CHANGES_REQUESTED** [review]

File: app/models/whitelisted_smart_contract.rb (hunk 6 above)
Comment (vpetrusenko) [reply app/models/whitelisted_smart_contract.rb]: Alex ask me to expose it, but sure i can remove it

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -99,6 +99,7 @@
     %w[101 102 201 202 211 212 301 302 401 402].each { |ac_code| FactoryBot.create(:operations_account, ac_code)}
     FactoryBot.create(:trading_fee, market_id: :any, group: :any, maker: 0.0015, taker: 0.0015)
     FactoryBot.create(:withdraw_limit)
+    %i[ address_1 address_2 address_3 address_4 ].each { |ccy| FactoryBot.create(:whitelisted_smart_contract, ccy) }
```
Comment (mnaichuk) [inline spec/spec_helper.rb]: don't need to create this in spec_helper

File: app/models/whitelisted_smart_contract.rb (hunk 6 above)
Comment (mnaichuk) [reply app/models/whitelisted_smart_contract.rb]: yes, please remove this, Alex agree

**Comment by mnaichuk** [comment]
Also please add a callback on smart contract creation to update `updated_at` of his blockhain

**Comment by vpetrusenko** [comment]
@mnaichuk updated

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:description][:desc] }
+          optional :state,
+                   values: { value: %w[active disabled], message: 'admin.whitelistedsmartcontract.invalid_state' },
+                   default: 'active',
+                   desc: -> { API::V2::Admin::Entities::WhitelistedSmartContract.documentation[:state][:desc] }
+        end
+        post '/whitelisted_smart_contracts' do
+          admin_authorize! :create, WhitelistedSmartContract
+
+          whitelisted_address = ::WhitelistedSmartContract.create!(declared(params))
```
Comment (chumaknadya) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: Create! will raise an error, so your save will not be executed

You can use .new and .save

File: app/models/whitelisted_smart_contract.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  # == Validations ==========================================================
+
+  validates :address, presence: true, uniqueness: { scope: :blockchain_key }
+
+  validates :blockchain_key,
+            presence: true,
+            inclusion: { in: ->(_) { Blockchain.pluck(:key).map(&:to_s) } }
+
+  validates :state,  inclusion: { in: %w[active disabled] }
+
```
Comment (chumaknadya) [inline app/models/whitelisted_smart_contract.rb:22]: Can you do frozen array of states here ?
And here just use this const, so it will make it reusable

**Review by chumaknadya — CHANGES_REQUESTED** [review]

**Comment by mnaichuk** [comment]
Also, we need to have the ability to import contacts from seeds please check the import rake task and add this functionality there

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+          if whitelisted_address.save
+            present whitelisted_address, with: API::V2::Admin::Entities::WhitelistedSmartContract
+            status 201
```
Comment (chumaknadya) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: you can do without status here

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        end
+        params do
+          requires :file,
+                   type: File
+        end
+        post '/whitelisted_smart_contracts/csv' do
+          admin_authorize! :create, WhitelistedSmartContract
+          count = 0
+
+          CSV.parse(params[:file][:tempfile], headers: true, quote_empty: false).each do |row|
```
Comment (chumaknadya) [inline app/api/v2/admin/whitelisted_smart_contracts.rb:120]: will it be always :tempfile variable here ?

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+            row = row.to_h.compact.symbolize_keys!
+            address = row[:address]
+            blockchain_key = row[:blockchain_key]
+            description = row[:description]
+            next if address.blank? || blockchain_key.blank? || !::Blockchain.pluck(:key).include?(blockchain_key)
```
Comment (chumaknadya) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: ```::Blockchain.pluck(:key).exclude?(blockchain_key)``` insead of ```!::Blockchain.pluck(:key).include?(blockchain_key)```

File: app/api/v2/admin/whitelisted_smart_contracts.rb [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+            status 422
+          end
+        end
+
+        desc 'Process whitelisted smart contracts from csv' do
+          success API::V2::Admin::Entities::WhitelistedSmartContract
+        end
+        params do
+          requires :file,
+                   type: File
```
Comment (chumaknadya) [inline app/api/v2/admin/whitelisted_smart_contracts.rb]: put description

File: spec/api/v2/admin/whitelisted_smart_contracts_spec.rb [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+
+describe API::V2::Admin::WhitelistedSmartContracts, type: :request do
+  let(:admin) { create(:member, :admin, :level_3, email: 'example15@gmail.com', uid: 'ID73BF61C8H1') }
+  let(:token) { jwt_for(admin) }
+  let(:level_3_member) { create(:member, :level_3) }
+  let(:level_3_member_token) { jwt_for(level_3_member) }
+  let(:test_file) { Rack::Test::UploadedFile.new(Rails.root.join('spec', 'resources', 'whitelisted_addresses', file_name), 'text/csv') }
+  let(:file_name) { '1.csv' }
+
+  let!(:addresses_1) { FactoryBot.create(:whitelisted_smart_contract, :address_1) }
```
Comment (chumaknadya) [inline spec/api/v2/admin/whitelisted_smart_contracts_spec.rb]: you can use just create(: whitelisted_smart_contract)

**Review by chumaknadya — CHANGES_REQUESTED** [review]

File: app/api/v2/admin/whitelisted_smart_contracts.rb (hunk 14 above)
Comment (vpetrusenko) [reply app/api/v2/admin/whitelisted_smart_contracts.rb:120]: should be, i write it with airdrop csv parse logic as example

**Comment by vpetrusenko** [comment]
@vpetrusenko @chumaknadya updated

**Pull request closed without merging**